### PR TITLE
[WIP] Reflect projects in /etc on kubelet restart

### PR DESCRIPTION
**Pull request by irajdeep**
As already mentioned in this issue: https://github.com/kubernetes/kubernetes/issues/79440, this change is to record projects in /etc
on kubelet restart so that we can prevent the case when other actors on the node that use project quota and rely on /etc may have collision with kubelet's project quotas

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR is a bug fix. As already mentioned in this [this issue](https://github.com/kubernetes/kubernetes/issues/79440). We would like to record  projects in /etc on kubelet restart so that other actors on the node that use project quota and rely on /etc, they may have collision with kubelet's project quotas
**Which issue(s) this PR fixes**:

Fixes # https://github.com/kubernetes/kubernetes/issues/79440

**Special notes for your reviewer**:

Hey folks, I created this PR as a draft, so that we have something to start a discussion on since I am not quite sure if the changes are headed in the right direction. Would be great if you can provide some initial feedback on this.

On a side note, I think the new function that I added `RecordProjectIDOnProjectFiles` is pretty inefficient, since it writes each existing project entry one by one into the files(which is quite an expensive operation), ideally, we would like to do it in a batch after collecting all the existing projects I reckon. For doing this we might need to add a new method in `project.go` (I don't see any existing method which does that). So providing some feedback here might also be helpful. 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

File: pkg/volume/util/fsquota/quota_linux.go [hunk 1]
```diff
@@ -439,3 +445,24 @@ func ClearQuota(m mount.Interface, path string) error {
 	}
 	return nil
 }
+
+// RecordProjectIDOnProjectFiles -- Records project file for the given directory-path to /etc/projects and /etc/projectid
+func RecordProjectIDOnProjectFiles(podVolumeDir string) error {
```
Comment (RobertKrawitz) [inline pkg/volume/util/fsquota/quota_linux.go]: This needs to be part of the interface, and a null version implemented in `quota.go`.  Otherwise it won't build on non-Linux systems.

Also make sure that this works with persistent /etc.

File: pkg/volume/util/fsquota/quota_linux.go [hunk 2]
```diff
@@ -220,6 +219,13 @@ func getQuotaOnDir(m mount.Interface, path string) (common.QuotaID, error) {
 	return getApplier(path).GetQuotaOnDir(path)
 }
 
+func getQuotaIDFromDirPath(path string) (common.QuotaID, bool) {
+	if val, ok := dirQuotaMap[path]; ok {
```
Comment (RobertKrawitz) [inline pkg/volume/util/fsquota/quota_linux.go]: `dirQuotaMap` won't be populated at kubelet start -- it only gets populated by `assignQuota`.  You need to use `GetQuotaOnDir` here.

File: pkg/volume/util/fsquota/quota_linux.go [hunk 3]
```diff
@@ -439,3 +445,24 @@ func ClearQuota(m mount.Interface, path string) error {
 	}
 	return nil
 }
+
+// RecordProjectIDOnProjectFiles -- Records project file for the given directory-path to /etc/projects and /etc/projectid
+func RecordProjectIDOnProjectFiles(podVolumeDir string) error {
+	if id, ok := getQuotaIDFromDirPath(podVolumeDir); ok {
+		fProjects, fProjid, err := openAndLockProjectFiles()
```
Comment (RobertKrawitz) [inline pkg/volume/util/fsquota/quota_linux.go]: You should be able to use `createProjectID` for this.

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -33,7 +33,7 @@ import (
 	"time"
 
 	cadvisorapi "github.com/google/cadvisor/info/v1"
-	v1 "k8s.io/api/core/v1"
```
Comment (RobertKrawitz) [inline pkg/kubelet/kubelet.go]: Why is this needed?

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -51,7 +51,7 @@ import (
 	"k8s.io/client-go/tools/record"
 	"k8s.io/client-go/util/certificate"
 	"k8s.io/client-go/util/flowcontrol"
-	cloudprovider "k8s.io/cloud-provider"
```
Comment (RobertKrawitz) [inline pkg/kubelet/kubelet.go]: Why is this needed?

**Review by RobertKrawitz — CHANGES_REQUESTED** [review]

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (irajdeep) [reply pkg/kubelet/kubelet.go]: My local `gofmt` seems to have done those automagically. Undid these changes

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (irajdeep) [reply pkg/kubelet/kubelet.go]: same as above

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (RobertKrawitz) [reply pkg/kubelet/kubelet.go]: If `hack/update-gofmt.sh` makes the changes that's OK.  But you shouldn't use your own `gofmt`.

File: pkg/volume/util/fsquota/quota_linux.go (hunk 1 above)
Comment (irajdeep) [reply pkg/volume/util/fsquota/quota_linux.go]: > This needs to be part of the interface, and a null version implemented in quota.go. Otherwise it won't build on non-Linux systems.

Done
> Also make sure that this works with persistent /etc.

Added a check for it. Want to know your thoughts on it: https://github.com/kubernetes/kubernetes/pull/82537/files#diff-1556842a9083b228aa95143da7468b79R468
If we need to write just in the file, we probably cannot use the `createProjectID` method and might instead need to use `writeprojectfile` method

File: pkg/volume/util/fsquota/quota_linux.go (hunk 3 above)
Comment (irajdeep) [reply pkg/volume/util/fsquota/quota_linux.go]: Done, also please check the comment above

File: pkg/volume/util/fsquota/quota_linux.go (hunk 2 above)
Comment (irajdeep) [reply pkg/volume/util/fsquota/quota_linux.go]: Done

File: pkg/volume/util/fsquota/quota_linux.go (hunk 1 above)
Comment (RobertKrawitz) [reply pkg/volume/util/fsquota/quota_linux.go]: Then perhaps `createProjectID` needs to be refactored.

Comment (irajdeep) [reply pkg/volume/util/fsquota/quota_linux.go]: How do you feel about using the `writeProjectFile` method instead?

**Comment by irajdeep** [comment]
Also, I was wondering how can we test these changes, I don't see unit test quite feasible here.
Integration test perhaps?

**Review by RobertKrawitz — CHANGES_REQUESTED** [review]

File: pkg/volume/util/fsquota/quota_linux.go [hunk 6]
```diff
@@ -439,3 +438,41 @@ func ClearQuota(m mount.Interface, path string) error {
...
+		if err != nil {
+			defer closeProjectFiles(fProjects, fProjid)
+			list := readProjectFiles(fProjects, fProjid)
+			projectList := list.projects
+			projectIDList := list.projid
+			foundInProjectList, foundInProjectIDList := false, false
+
+			for _, project := range projectList {
+				if project.id == id && project.data == podVolumeDir {
```
[the remark this answers is not in the document]
Comment (irajdeep) [reply pkg/volume/util/fsquota/quota_linux.go]: * The case in which project.data matches but not project.id, I am wondering if it's possible, because `GetQuotaOnDir` returns the current id applied to the directory.
* If the project.id matches but the project.data doesn't mean the current project.id is assigned to multiple directories, so we should add a new entry in the projectList for the current `id` and `podVolumeDir` which is happening in the current implementation.

Am I missing some case here?
